cmd/clusterctl/client/cluster/objectgraph.go

```diff
@@ -615,14 +669,46 @@ func (o *objectGraph) setSoftOwnership() {
 	}
 
 	clusterClasses := o.getClusterClasses()
+
+	// Collect all cross-ns ClusterClasses to exclude from deletion
+	for _, clusterClass := range clusterClasses {
+		for _, cluster := range o.externalObjects {
+			if cluster.identity.GroupVersionKind().GroupKind() != clusterv1.GroupVersion.WithKind("Cluster").GroupKind() {
+				continue
+			}
+
+			// if the cluster uses a managed topology and uses the clusterclass
+			// check the namespace of the class
+			className, hasName := cluster.additionalInfo[clusterTopologyNameKey]
+			classNamespace, hasNamespace := cluster.additionalInfo[clusterTopologyNamespaceKey]
+			if namespace, ok := classNamespace.(string); !hasNamespace || !ok {
+				classNamespace = cmp.Or(namespace, cluster.identity.Namespace)
+			}
+
+			// Set shouldNotDelete if clusterClass already referenced by some external cluster
+			if hasName && className == clusterClass.identity.Name && clusterClass.identity.Namespace == classNamespace {
```

Comment:
If I got it right, externalObjects contains all the clusters also the ones in the namespace being moved.

If this is true, then we are preventing deleting also for CC referenced only from clusters in the namespace being moved which is something we don't want (it will be a breaking change).

Reply:
No, those are actually excluded at the beginning of `addExternal` by matching against UID of objects set to move.